protocols/request-response/src/lib.rs

```diff
@@ -571,12 +624,13 @@ where
     fn inject_event(
         &mut self,
         peer: PeerId,
-        _: ConnectionId,
+        connection: ConnectionId,
         event: RequestResponseHandlerEvent<TCodec>,
     ) {
         match event {
             RequestResponseHandlerEvent::Response { request_id, response } => {
-                self.pending_responses.remove(&request_id);
+                self.remove_pending_inbound_response(&peer, connection, &request_id);
```

Comment:
So the assumption that justifies panicking in `remove_pending_inbound_response` and `remove_pending_outbound_response` if the `request_id` is absent is that `inject_event` is never expected to be called with a `ConnectionId` after `inject_connection_closed` has already been called for that same connection, right? Wouldn't it be a bit clearer to move the `debug_assert!`s to the call-sites here, clarifying the expectations there and letting these `remove` methods return a `bool`, avoiding the extra `bool` parameter for `remove_pending_outbound_response`? Telling a method whether it should make a debug assertion or not seems a bit weird.

Reply:
`remove_pending_inbound_response` and `remove_pending_outbound_response` only panic in debug mode if the `request_id` is absent AND the connection is still alive. If the connection is not alive anymore, i.e. `inject_connection_closed` was called before `inject_event`, `remove_pending_inbound_response` and `remove_pending_outbound_response` do not panic when the `request_id` is absent.

> `inject_event` is never expected to be called with a `ConnectionId` after `inject_connection_closed` has already been called for that same connection, right?

At the time of writing the two functions I was not sure this is a guarantee the `Network` provides. Taking another look it does seem to be the case. Is this a guarantee that we want or can uphold in the future @romanb? If so I think your suggestion above of moving the `debug_assert` to the caller side simplifies the logic.

Reply:
> Taking another look it does seem to be the case. Is this a guarantee that we want or can uphold in the future @romanb?

Yes, I think that `ConnectionClosed` is the last event emitted for a connection (and hence `inject_connection_closed` being invoked) is an elementary property.

Reply:
Great. As said above this allows some simplifications done with https://github.com/libp2p/rust-libp2p/pull/1886/commits/bb310990a7bd3cbc46a8d8e1f67450d69c6fc899. Thanks for thinking this through!

As an aside I am not 100 % happy with the use of `debug_assert` in some and `expect` in other places. I would be fine with replacing these `debug_assert` calls with `expect` calls, but I am not sure which _panic-policy_ you have been following within rust-libp2p.

Reply:
My only panic-policy is "Don't panic" :lotus_position:. I think `expect` are stronger invariants where execution cannot continue if they are violated, whereas with `debug_assert!`s that is not necessarily the case, even though they almost always indicate a bug when violated, of course.